source/utilities.cc

```diff
@@ -2535,7 +2535,67 @@ namespace aspect
       return operator_list;
     }
 
+    template <int dim>
+    GeneralizedPoint<dim>::GeneralizedPoint(Point<dim> position, const GeometryModel::Interface<dim> *geometry_model)
+    {
+      cartesian_position = position;
+      coordinate_system = geometry_model->natural_coordinate_system();
+      coordinates = geometry_model->cartesian_to_natural_coordinates(cartesian_position);
+    }
+
+    template <int dim>
+    Point<dim> &GeneralizedPoint<dim>::get_position()
+    {
+      return cartesian_position;
+    }
+
+    template <int dim>
+    std_cxx11::array<double,dim> &GeneralizedPoint<dim>::get_coordinates()
+    {
+      return coordinates;
+    }
+
+    template <int dim>
+    std_cxx11::array<double,dim-1> GeneralizedPoint<dim>::get_surface_coordinates() const
+    {
+      if (coordinate_system == Coordinates::CoordinateSystem::cartesian)
+        {
+          if (dim == 2)
+            {
+              std_cxx11::array<double,1> a {coordinates[0]};
+              return reinterpret_cast<std_cxx11::array<double,dim-1>& >(a);
+            }
+          else if (dim == 3)
+            {
+              std_cxx11::array<double,2> a {coordinates[0],coordinates[1]};
+              return reinterpret_cast<std_cxx11::array<double,dim-1>& >(a);
+            }
+        }
+      else if (coordinate_system == Coordinates::CoordinateSystem::spherical || coordinate_system == Coordinates::CoordinateSystem::ellipsoidal)
+        {
+          if (dim == 2)
+            {
+              std_cxx11::array<double,1> a {coordinates[1]};
+              return reinterpret_cast<std_cxx11::array<double,dim-1>& >(a);
+            }
+          else if (dim == 3)
+            {
+              std_cxx11::array<double,2> a {coordinates[1],coordinates[2]};
+              return reinterpret_cast<std_cxx11::array<double,dim-1>& >(a);
+            }
+        }
+      return std_cxx11::array<double,dim-1> {};
+    }
 
+    template <int dim>
+    double GeneralizedPoint<dim>::get_depth_coordinate() const
+    {
+      if (coordinate_system == Coordinates::CoordinateSystem::cartesian)
+        return coordinates[dim-1];
+      else if (coordinate_system == Coordinates::CoordinateSystem::spherical || coordinate_system == Coordinates::CoordinateSystem::ellipsoidal)
+        return coordinates[0];
+      return 0;
```

Comment:
same here -- with a `switch` statement, things become a bit simpler, and you can also catch the case that the `coordinate_system` has a value that has not been checked before.